x-pack/filebeat/processors/decode_cef/cef/cef_actions.rl

```diff
@@ -0,0 +1,88 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+%%{
+    machine cef;
+
+    # Actions to execute while executing state machine.
+    action mark {
+        mark = p
+    }
+    action mark_slash {
+        mark_slash = p
+    }
+    action mark_escape {
+        state.pushEscape(mark_slash, p)
+    }
+    action version {
+        e.Version, _ = strconv.Atoi(data[mark:p])
+    }
+    action device_vendor {
+        e.DeviceVendor = replaceEscapes(data[mark:p], mark, state.escapes)
+        state.reset()
+    }
+    action device_product {
+        e.DeviceProduct = replaceEscapes(data[mark:p], mark, state.escapes)
+        state.reset()
+    }
+    action device_version {
+        e.DeviceVersion = replaceEscapes(data[mark:p], mark, state.escapes)
+        state.reset()
+    }
+    action device_event_class_id {
+        e.DeviceEventClassID = replaceEscapes(data[mark:p], mark, state.escapes)
+        state.reset()
+    }
+    action name {
+        e.Name = replaceEscapes(data[mark:p], mark, state.escapes)
+        state.reset()
+    }
+    action severity {
+        e.Severity = data[mark:p]
+    }
+    action complete_header {
+        complete = true
+    }
+    action incomplete_header {
+        mark = p
+        state.reset()
+    }
+    action extension_key {
+        // A new extension key marks the end of the last extension value.
+        if len(state.key) != 0 && state.valueStart < mark {
+            // We should not be here, but purge the escapes and handle them.
+            e.pushExtension(state.key, replaceEscapes(data[state.valueStart:mark-1], state.valueStart, state.escapes))
+            state.reset()
+        }
+        state.key = data[mark:p]
+    }
+    action extension_value_start {
+        if len(state.escapes) != 0 {
```

Comment:
Why is this condition based on the number of escapes? Can't we have extensions without escapes?

Reply:
The issue here that if we have got to this point we have a syntax that is not expected and so have not consumed the escapes. So if there are some to consume, this makes sure we do.

Reply:
My point is that there may be an "extension" that needs to be consumed which does not contain any escapes. That's why I am questioning the condition being based on len(escapes) rather than something like len(key) && something with the value start/end. Does that make sense?

Reply:
Those won't be affected by this. This is purely to cope with the cases where things have been set up and not used. I the first pass of the parse, this condition is never triggered because the header state machine prevents this from ever happening, but in the salvage pass, the guarantees don't hold, but also we can't guarantee sensible results, we are just trying to capture as much as possible.

Put another way, if the escape were well formed, it would have been handled by `extension_eof` which is those cases that you describe, guarded by `len(state.key) != 0 && state.valueStart < state.valueEnd`. So this just picks up cases that that failed to identify because they were syntactically broken and did not satisfy the machine before the next extension start.

Does that clarify?

Reply:
Yes, that’s clear. Thanks. Can you add a short comment there.